build_system/clerk.ml

```diff
@@ -159,104 +179,273 @@ let search_for_expected_outputs (file : string) : expected_output_descr list =
 
 type testing_result = { error_code : int; number_of_tests_run : int; number_correct : int }
 
-let test_file (tested_file : string) (catala_exe : string) (catala_opts : string)
-    (reset_test_outputs : bool) : testing_result =
+let add_reset_rules_aux ~(redirect : string) ~(with_scope_output_rule : string)
+    ~(without_scope_output_rule : string) (catala_exe_opts : string) (rules : Rule.t Nj.RuleMap.t) :
+    Rule.t Nj.RuleMap.t =
+  let reset_common_cmd_exprs =
+    Nj.Expr.
+      [
+        Var "catala_cmd";
+        Var "tested_file";
+        Var "extra_flags";
+        Lit "--unstyled";
+        Lit redirect;
+        Var "expected_output";
+        Lit "2>&1";
+      ]
+  in
+  let reset_with_scope_rule =
+    Nj.Rule.make with_scope_output_rule
+      ~command:
+        Nj.Expr.(Seq ([ Lit catala_exe_opts; Lit "-s"; Var "scope" ] @ reset_common_cmd_exprs))
+      ~description:
+        Nj.Expr.(
+          Seq
+            [
+              Lit "RESET scope";
+              Var "scope";
+              Lit "of file";
+              Var "tested_file";
+              Lit "with the";
+              Var "catala_cmd";
+              Lit "command";
+            ])
+  in
+  let reset_without_scope_rule =
+    Nj.Rule.make without_scope_output_rule
+      ~command:Nj.Expr.(Seq (Lit catala_exe_opts :: reset_common_cmd_exprs))
+      ~description:
+        Nj.Expr.(
+          Seq
+            [
+              Lit "RESET";
+              Lit "file";
+              Var "tested_file";
+              Lit "with the";
+              Var "catala_cmd";
+              Lit "command";
+            ])
+  in
+  Nj.RuleMap.(
+    rules
+    |> add reset_with_scope_rule.name reset_with_scope_rule
+    |> add reset_without_scope_rule.name reset_without_scope_rule)
+
+let add_test_rules_aux ~(test_common_cmd_exprs : Expr.t list) ~(with_scope_output_rule : string)
+    ~(without_scope_output_rule : string) (catala_exe_opts : string) (rules : Rule.t Nj.RuleMap.t) :
+    Rule.t Nj.RuleMap.t =
+  let test_with_scope_rule =
+    Nj.Rule.make with_scope_output_rule
+      ~command:
+        Nj.Expr.(Seq ([ Lit catala_exe_opts; Lit "-s"; Var "scope" ] @ test_common_cmd_exprs))
+      ~description:
+        Nj.Expr.(
+          Seq
```

Comment:
Splendid! I just wanted to say that generally your code is admirably clean and readable.